src/ukkonen.cpp

```diff
@@ -0,0 +1,936 @@
+
+//
+// libsemigroups_pybind11
+// Copyright (C) 2024 Joseph Edwards
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// libsemigroups headers
+#include <libsemigroups/types.hpp>  // for word_type
+#include <libsemigroups/ukkonen.hpp>
+
+// pybind11....
+#include <pybind11/operators.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+// libsemigroups_pybind11....
+#include "main.hpp"  // for init_ukkonen
+
+namespace py = pybind11;
+
+namespace libsemigroups {
+
+  template <typename Word>
+  void bind_ukkonen_extras(py::module& m, py::class_<Ukkonen>& uk) {
+    ////////////////////////////////////////////////////////////////////////
+    // Ukkonen
+    ////////////////////////////////////////////////////////////////////////
+
+    uk.def(
+        "index",
+        [](Ukkonen const& self, Word const& w) {
+          return self.index(w.begin(), w.end());
+        },
+        py::arg("w"),
+        R"pbdoc(
+:sig=(self: Ukkonen, w: str|List[int]) -> int:
+:only-document-once:
+
+Find the index of a word in the suffix tree.
+
+If the *w* is one of the words that the suffix tree contains
+(the words added to the suffix tree via :any:`ukkonen.add_word`), then this function
+returns the index of that word. If the word *w* is not one of the words that the
+suffix tree represents, then :any:`UNDEFINED` is returned.
+
+:param w: the word to check.
+:type w: str|List[int]
```

Comment:
No idea if this impacts the generated doc or not but isn't there some missing whitespace here?

Reply:
Between `str` and `list`?

Reply:
Yup